sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -358,6 +360,35 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   }
 
   /**
+   * :: Experimental ::
+   * Starts the execution of the streaming query, which will continually send results to the given
+   * [[ForeachWriter]] as new data arrives. The returned [[ContinuousQuery]] object can be used to
+   * interact with the stream.
+   *
+   * @since 2.0.0
+   */
+  @Experimental
+  def foreach(writer: ForeachWriter[T]): ContinuousQuery = {
+    val queryName = extraOptions.getOrElse("queryName", StreamExecution.nextName)
+    val checkpointLocation = extraOptions.get("checkpointLocation")
+      .orElse {
+        df.sparkSession.sessionState.conf.checkpointLocation.map { l =>
+          new Path(l, queryName).toUri.toString
+        }
+      }.getOrElse {
```

Comment:
I think similar to memorysink, the user should not have to specify checkpoint location. 

Also could you consolidate the code for finding checkpoint location across memory sink, file sink and foreach sink.